packages/Symfony/src/Rector/Class_/MakeCommandLazyRector.php

```diff
@@ -0,0 +1,155 @@
+
+namespace Rector\Symfony\Rector\Class_;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\Node\Stmt\Expression;
+use Rector\PhpParser\NodeTraverser\CallableNodeTraverser;
+use Rector\Rector\AbstractRector;
+use Rector\RectorDefinition\CodeSample;
+use Rector\RectorDefinition\RectorDefinition;
+
+/**
+ * @see https://symfony.com/doc/current/console/commands_as_services.html
+ */
+final class MakeCommandLazyRector extends AbstractRector
+{
+    /**
+     * @var string
+     */
+    private const COMMAND_CLASS = 'Symfony\Component\Console\Command\Command';
+
+    /**
+     * @var CallableNodeTraverser
+     */
+    private $callableNodeTraverser;
+
+    public function __construct(CallableNodeTraverser $callableNodeTraverser)
+    {
+        $this->callableNodeTraverser = $callableNodeTraverser;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Make Symfony commands lazy', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+use Symfony\Component\Console\Command\Command
+
+class SunshineCommand extends Command
+{
+    public function configure()
+    {
+        $this->setName('sunshine');
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+use Symfony\Component\Console\Command\Command
+
+class SunshineCommand extends Command
+{
+    protected static $defaultName = 'sunshine';
+    public function configure()
+    {
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $this->isType($node, self::COMMAND_CLASS)) {
+            return null;
+        }
+
+        $commandName = $this->resolveCommandNameAndRemove($node);
+        if ($commandName === null) {
+            return null;
+        }
+
+        $defaultNameProperty = $this->createDefaultNameProperty($commandName);
+        $node->stmts = array_merge([$defaultNameProperty], (array) $node->stmts);
+
+        return $node;
+    }
+
+    private function createDefaultNameProperty(Node $commandNameNode): Node\Stmt\Property
+    {
+        return $this->builderFactory->property('defaultName')
+            ->makeProtected()
+            ->makeStatic()
+            ->setDefault($commandNameNode)
+            ->getNode();
+    }
+
+    private function resolveCommandNameAndRemove(Class_ $class): ?Node
+    {
+        $commandName = null;
+        $this->callableNodeTraverser->traverseNodesWithCallable((array) $class->stmts, function (Node $node) use (
+            &$commandName
+        ) {
+            if ($node instanceof MethodCall) {
+                if (! $this->isName($node, 'setName')) {
```

Comment:
Done testing & this is too fragile, you must check it's call on `$this` (or fluent) cause I saw replacement:
```
+    protected static $defaultName = $something->name;
-                $object->setName($something->name);

Reply:
How would you solve that?

Reply:
Can we check on what the method call is being done? If so I would validate its object called `$this` or of command instance.

Reply:
Yes, that's already done.

 
The problem is with "($something->name)". It cannot be move to static property.

Reply:
> The problem is with "($something->name)". It cannot be move to static property.

Yes, but this call is not on the command itself, it's on totally unrelated object (in this case it's value object class :) )

Reply:
Ok, test class will be much cleaner I guess :)
```php
class SomeCommand extends ContainerAwareCommand
{
    protected function configure()
    {
        $this->setDescription('Description');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
         // This will fail
         $input->setName('whatever you put here, object, string, you name it');
    }
}
```
Second:
```
class SomeCommand extends ContainerAwareCommand
{
    protected function configure()
    {
        $this->setName('some:command');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
         // This will fail
         $input->setName('whatever you put here, object, string, you name it');
    }
}